pkg/operator/ceph/object/notification/controller.go

```diff
@@ -0,0 +1,240 @@
+		return errors.Wrapf(err, "failed to retrieve ObjectBucket %q", cephOBC.Spec.ObjectBucketName)
+	}
+	cephSC := storagev1.StorageClass{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: cephOBC.Spec.StorageClassName}, &cephSC); err != nil {
+		return errors.Wrapf(err, "failed to retrieve StorageClass %q", cephOBC.Spec.StorageClassName)
+	}
+
+	// step 2: delete all existing notifications
+	p := Provisioner{
+		Client:      r.client,
+		Context:     r.context,
+		ClusterInfo: r.clusterInfo,
+		ClusterSpec: r.clusterSpec,
+	}
+	if err := p.CreateSession(cephOB.Spec.AdditionalState["cephUser"], cephSC.Parameters["objectStoreName"]); err != nil {
+		return errors.Wrapf(err, "failed to create session for bucket notification. bucket: %q", cephOB.Name)
+	}
+	if !p.HasSession() {
+		return errors.Errorf("failed to create session for bucket notification. bucket: %q", cephOB.Name)
+	}
+	err := p.DeleteAll(&cephOB)
+	if err != nil {
+		return errors.Wrapf(err, "failed delete all bucket notification from bucket: %q", cephOB.Name)
+	}
+
+	// Step 3: looking for notifications in the labels
+	for key, name := range cephOBC.Labels {
+		notifyLabels := strings.SplitAfterN(key, notificationLabelPrefix, 2)
+		if len(notifyLabels) > 1 && notifyLabels[1] != "" && name == notifyLabels[1] {
+			// Step 4: for each notification label fetch the bucket notification CRD
+			logger.Debugf("bucket notification label: %q found on ObjectbucketClaim: %q", name, cephOBC.Name)
+			bucketNotification := &cephv1.CephBucketNotification{}
+			if err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: name}, bucketNotification); err != nil {
+				return errors.Wrapf(err, "failed to retrieve CephBucketNotification %q", name)
+			}
+			logger.Debugf("CephBucketNotification: %q resource found", bucketNotification.Name)
+
+			// Step 5: verify that the associated topic configuration exists with an ARN
+			bucketTopic := &cephv1.CephBucketTopic{}
+			if err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: namespace, Name: bucketNotification.Spec.Topic}, bucketTopic); err != nil {
+				return errors.Wrapf(err, "failed to retrieve CephBucketTopic %q", name)
+			}
+			topicARN := topic.GetTopicARN(bucketTopic)
+			if topicARN == "" {
+				return errors.New("no ARN in topic. topic was not provisioned successfully")
+			}
+			parsedTopicARN, err := arn.Parse(topicARN)
+			if err != nil {
+				return err
+			}
+			if strings.ToLower(parsedTopicARN.Service) != "sns" {
+				return errors.New("topic ARN must have 'sns' service")
+			}
+			if parsedTopicARN.Resource == "" {
+				return errors.New("missing topic inside ARN")
+			}
+			logger.Debugf("CephBucketTopic found with valid ARN: %q", topicARN)
+
+			// step 6: provision the notification
+			err = p.Create(&cephOB, topicARN, bucketNotification)
```

Comment:
What if the `bucketNotification` already  exists for OBC, do PutBucketNotificationConfiguration() fails?
I mean how the below situation handled, Initially I add one `bucket-notifcation-1` to OBC. Later I add `bucket-notifcation-2`. So while handling `bucket-notification-2`, do I ended up creating  `bucket-notifcation-1` again?

Reply:
* the first action is to delete all notifications of the bucket, so we have to create all notification back again. we do this cleanup because we don't know which notifications were removed from the list of labels
* even if we didn't do this cleanup first, there is no problem with creating a notification that is already there. all operations we do on the RGW here are idempotent